sql/sql_delete.cc

```diff
@@ -254,7 +254,12 @@ int TABLE::delete_row()
 
   store_record(this, record[1]);
   vers_update_end();
-  return file->ha_update_row(record[1], record[0]);
+  int res;
+  if ((res= file->extra(HA_EXTRA_REMEMBER_POS)))
+    return res;
+  if ((res= file->ha_update_row(record[1], record[0])))
+    return res;
+  return file->extra(HA_EXTRA_RESTORE_POS);
 }
```

Comment:
Hmm, I don't know... If I understand correctly, this bug is happening basically because you're doing an update, not a delete. But you pretend you're doing a delete, so the server doesn't know what you're really doing and chooses the wrong execution plan. I mean, if there's no versioning, but a normal timestamp field `row_end`, and you're just doing `update t1 set row_end=now(6) where pk >= 0` then there will be no error, right?

So perhaps the correct fix is not to lie to the server in the first place and execute `mysql_update`, not `mysql_delete`? Instead of pretending there's a delete and patching issues that will be arising from that. I'm not at all sure this one will be the last.

Reply:
> Hmm, I don't know... If I understand correctly, this bug is happening basically because you're doing an update, not a delete. But you pretend you're doing a delete, so the server doesn't know what you're really doing and chooses the wrong execution plan. I mean, if there's no versioning, but a normal timestamp field row_end, and you're just doing update t1 set row_end=now(6) where pk >= 0 then there will be no error, right?

Right, no error. I wasn't able to reproduce an error without system versioning.

> So perhaps the correct fix is not to lie to the server in the first place and execute mysql_update, not mysql_delete? Instead of pretending there's a delete and patching issues that will be arising from that. I'm not at all sure this one will be the last.

There is also `multi_delete`. In other cases sounds reasonable indeed. But I don't know how would this change bug rate. I would expect other bugs, related to `mysql_updated` usage. My opinion was that this bug appears because `row_end` is a part of a `PRIMARY KEY`. See [here](https://github.com/tempesta-tech/mariadb/issues/482) our analysis.